util/src/using_queue.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+//! Queue-like datastructure including notion of usage.
+
+/// Special queue-like datastructure that includes the notion of
+/// usage to avoid items that were queued but never used from making it into
+/// the queue.
+pub struct UsingQueue<T> where T: Clone {
+	/// Not yet being sealed by a miner, but if one asks for work, we'd prefer they do this.
+	pending: Option<T>,
+	/// Currently being sealed by miners.
+	in_use: Vec<T>,
```

Comment:
Functions accepting predicates (below) are always returning only single value. So I guess it's possible to just fetch using some key. `LinkedHashMap` would be perfect here.
Something like this?
https://contain-rs.github.io/linked-hash-map/linked_hash_map/struct.LinkedHashMap.html